schema/shared/src/main/scala-3/zio/blocks/schema/DerivedOptics.scala

```diff
@@ -0,0 +1,199 @@
+package zio.blocks.schema
+
+import scala.quoted.*
+import scala.language.dynamics
+
+/**
+ * A trait that can be extended by companion objects to automatically derive
+ * optics (Lens for case class fields, Prism for sealed trait/enum variants).
+ *
+ * Usage:
+ * {{{
+ * case class Person(name: String, age: Int)
+ * object Person extends DerivedOptics[Person]
+ *
+ * // Access optics via the `optics` member:
+ * val nameLens: Lens[Person, String] = Person.optics.name
+ * val ageLens: Lens[Person, Int] = Person.optics.age
+ * }}}
+ *
+ * For sealed traits/enums:
+ * {{{
+ * sealed trait Shape
+ * case class Circle(radius: Double) extends Shape
+ * case class Rectangle(width: Double, height: Double) extends Shape
+ * object Shape extends DerivedOptics[Shape]
+ *
+ * // Access prisms via the `optics` member:
+ * val circlePrism: Prism[Shape, Circle] = Shape.optics.Circle
+ * }}}
+ *
+ * The optics object is cached to avoid recreation on every access.
+ *
+ * @tparam S
+ *   The type for which to derive optics
+ */
+trait DerivedOptics[S] {
+
+  /**
+   * Provides access to the derived optics for type S. For case classes, returns
+   * an object with Lens accessors for each field. For sealed traits/enums,
+   * returns an object with Prism accessors for each variant.
+   *
+   * The returned object uses structural typing, so you get compile-time type
+   * checking and IDE completion for the accessor names.
+   */
+  transparent inline def optics(using schema: Schema[S]): Any = ${ DerivedOpticsMacros.opticsImpl[S]('schema) }
+}
+
+/**
+ * An optics holder that stores lenses/prisms in a map and provides dynamic
+ * access.
+ */
+final class OpticsHolder(members: Map[String, Any]) extends scala.Dynamic with Selectable {
+  def selectDynamic(name: String): Any =
+    members.getOrElse(name, throw new RuntimeException(s"No optic found for: $name"))
+}
+
+private object DerivedOpticsMacros {
+  import java.util.concurrent.ConcurrentHashMap
+
+  // Global cache to avoid recreating optics objects at runtime
+  // Key is the type's full name as a string
+  private val cache = new ConcurrentHashMap[String, OpticsHolder]()
+
+  def getOrCreate(key: String, create: => OpticsHolder): OpticsHolder = {
+    var result = cache.get(key)
+    if (result == null) {
+      result = create
+      val existing = cache.putIfAbsent(key, result)
+      if (existing != null) result = existing
+    }
+    result
+  }
+
+  def opticsImpl[S: Type](schema: Expr[Schema[S]])(using q: Quotes): Expr[Any] = {
+    import q.reflect.*
+
+    val caseClassType = TypeRepr.of[S]
+    val caseClassSym  = caseClassType.typeSymbol
+
+    // Check if it's a case class or sealed trait
+    val isCaseClass = caseClassSym.flags.is(Flags.Case)
+    val isSealed    = caseClassSym.flags.is(Flags.Sealed)
```

Comment:
I haven't verified the code, but note that Scala 3 `enum` should count the same as `sealed trait`.

You should have automated tests to verify this.